stdlib/public/runtime/HeapObject.cpp

```diff
@@ -78,6 +79,32 @@ HeapObject *swift::swift_allocObject(HeapMetadata const *metadata,
   return _swift_allocObject(metadata, requiredSize, requiredAlignmentMask);
 }
 
+#if OBJC_SETASSOCIATEDOBJECTHOOK_DEFINED
+//We interpose objc_setAssociatedObject so that we can set a flag in
+//the refcount field of Swift objects to indicate that they have associations,
+//since we can't safely skip ObjC dealloc work if they do
+static objc_hook_setAssociatedObject originalAssocObjectFunc = nullptr;
+
+static void _swift_setAssociatedObject_hook(
+  id _Nonnull object,
+  const void * _Nonnull key,
+  id _Nullable value,
+  objc_AssociationPolicy policy
+) {
+  if (!isObjCTaggedPointerOrNull(object) &&
+      objectUsesNativeSwiftReferenceCounting(object)) {
+    auto heapObj = reinterpret_cast<HeapObject *>(object);
+    heapObj->refCounts.setHasNoObjCComplications(false);
```

Comment:
This double negative makes me a bit sad but I can't quite think of a better way to phrase it.

Reply:
"can be deallocated without ObjC"?

Reply:
`HasPureSwiftDestruct` maybe. Or invert the meaning and call it `RequiresObjCDestruct`.

Reply:
I assumed the non-inversion is deliberate, so that the 0 case retains its former meaning.

Reply:
Indeed. Also phrasing it this way makes the check in deallocation not inverted. HasPureSwiftDestruct would be reasonable though, I like that.